clcache/__main__.py

```diff
@@ -1527,46 +1530,78 @@ def createManifestEntry(manifestHash, includePaths):
 
 
 def main():
-    if len(sys.argv) == 2 and sys.argv[1] == "--help":
-        print("""
-clcache.py v{}
-  --help    : show this help
-  -s        : print cache statistics
-  -c        : clean cache
-  -C        : clear cache
-  -z        : reset cache statistics
-  -M <size> : set maximum cache size (in bytes)
-""".strip().format(VERSION))
-        return 0
+    # These Argparse Actions are necessary because the first commandline
+    # argument, the compiler executable path, is optional, and the argparse
+    # class does not support conditional selection of positional arguments.
+    # Therefore, these classes check the candidate path, and if it is not an
+    # executable, stores it in the namespace as a special variable, and
+    # the compiler argument Action then prepends it to its list of arguments
+    class CommandCheckAction(argparse.Action):
+        def __call__(self, parser, namespace, values, optional_string=None):
+            if values and not values.endswith(".exe"):
+                setattr(namespace, "non_command", values)
+                return
+            setattr(namespace, self.dest, values)
+
+    class RemainderSetAction(argparse.Action):
+        def __call__(self, parser, namespace, values, optional_string=None):
+            nonCommand = getattr(namespace, "non_command", None)
+            if nonCommand:
+                values.insert(0, nonCommand)
+            setattr(namespace, self.dest, values)
+
+    parser = argparse.ArgumentParser(description="clcache.py v" + VERSION)
```

Comment:
I see that you scratched your own itch (#269 ), too. That's the spirit. :-)

I was quite skeptical about `argparse` actually working as intended, remembering the strange handling of quotation marks on the Windows command line. I seemed to recall that the argument parsing done by `argparse` was not bug-compatible (ahem) with what cl.exe ecpects. However, seeing that the tests still pass, maybe it's not so bad after all, phew! :-)

It doesn't seem to be a big improvement in terms of lines of code saved though, so I wonder -- any particular reason you started to use `argparse` here?

Reply:
We have been using argparse this way since I started using clcache, although the organization was a bit different; the combination into one is just a few days old. If there were any issues with processing cl (or other commands') command line arguments, aside from what we hit wit clang-cl (which wasn't an argparse issue) we would probably have seen it long ago.

Argparse may, in this case, have some issues with space compared to the old way, but it does provide a standard way of specifying parameters (including that the query options are mutually exclusive), datatypes, and help. Additionally, with this organization it may be possible to add extra command line arguments, in addition to the optional compiler path I added. It may also be possible to add processing of the compiler arguments (what is done in the command line analyzer) , but I haven't investigated that, not sure it is worth the extra work.

Reply:
Fair enough! I don't mind using `argparse` at all really (it just never seemed important enough for me to tackle it myself) as long as it doesn't break stuff. I was just curious whether maybe there's some other benefit I didn't notice.

The command line analyzer is a good point -- I think somebody _did_ try to use `argparser` at first but the noticed that it might be easier to reverse-engineer to peculiar command line handling of `cl.exe` manually, and that then became the command line analyzer.